File: android/app/src/main/java/io/mosip/residentapp/IntentData.java
package io.mosip.residentapp;

public class IntentData {
    private static final String FLOW_QR_LOGIN = "qrLoginFlow";
    private static final String FLOW_OVP = "ovpFlow";
    private static final String FLOW_CREDENTIAL_OFFER = "credentialOfferFlow";

    private String qrData = "";
    private String ovpQrData = "";
    private String credentialOfferData = "";

    private static IntentData intentData;
    public static IntentData getInstance() {
        if(intentData == null)
            intentData = new IntentData();
        return intentData;
    }
    public String getQrData() {
        return qrData;
    }

    public void setQrData(String qrData) {
        this.qrData = qrData;
    }

    public String getOVPQrData() {
        return ovpQrData;
    }

    public void setOVPQrData(String ovpQrData) {
        this.ovpQrData = ovpQrData;
    }

    // Read-once-and-clear: returning the URI also wipes the native bucket
    // so a re-entry of app.ready.focus.active cannot replay the same intent.
    public synchronized String getCredentialOfferData() {
        String data = this.credentialOfferData;
        this.credentialOfferData = "";
        return data;
    }

    public synchronized void setCredentialOfferData(String credentialOfferData) {
        this.credentialOfferData = credentialOfferData;
    }

    public String getDataByFlow(String flowType) {
        if (flowType == null) return "";
        return switch (flowType) {
            case FLOW_QR_LOGIN -> getQrData();
            case FLOW_OVP -> getOVPQrData();
            case FLOW_CREDENTIAL_OFFER -> getCredentialOfferData();
            default -> "";
        };
    }

    public void resetDataByFlow(String flowType) {
        if (flowType == null) return;
        switch (flowType) {
            case FLOW_QR_LOGIN -> setQrData("");
            case FLOW_OVP -> setOVPQrData("");
            case FLOW_CREDENTIAL_OFFER -> setCredentialOfferData("");
            default -> {
            }
        }
    }
}

File: ios/IntentData.swift
import Foundation

@objc public class IntentData: NSObject {
  @objc public static let shared = IntentData()
  private let syncQueue = DispatchQueue(label: "com.intentdata.syncQueue", attributes: .concurrent)
  private var qrData: String = ""
  private var ovpQrData: String = ""
  private var credentialOfferData: String = ""

  private override init() {
    super.init()
  }

  @objc public func getQrData() -> String {
    var data: String = ""
    syncQueue.sync {
      data = qrData
    }
    return data
  }

  @objc public func setQrData(_ newValue: String) {
    syncQueue.async(flags: .barrier) {
      self.qrData = newValue
    }
  }

  @objc public func getOvpQrData() -> String {
    var data: String = ""
    syncQueue.sync {
      data = ovpQrData
    }
    return data
  }

  @objc public func setOvpQrData(_ newValue: String) {
    syncQueue.async(flags: .barrier) {
      self.ovpQrData = newValue
    }
  }

  // Read-once-and-clear: returning the URI also wipes the native bucket
  // so a re-entry of app.ready.focus.active cannot replay the same intent.
  @objc public func getCredentialOfferData() -> String {
    var data: String = ""
    syncQueue.sync(flags: .barrier) {
      data = self.credentialOfferData
      self.credentialOfferData = ""
    }
    return data
  }

  @objc public func setCredentialOfferData(_ newValue: String) {
    syncQueue.async(flags: .barrier) {
      self.credentialOfferData = newValue
    }
  }

  func getDataByFlow(_ flowType: String?) -> String {
    switch flowType {
    case "qrLoginFlow":
      return getQrData()
    case "ovpFlow":
      return getOvpQrData()
    case "credentialOfferFlow":
      return getCredentialOfferData()
    default:
      return ""
    }
  }

  func resetDataByFlow(_ flowType: String) {
    switch flowType {
    case "qrLoginFlow":
      setQrData("")
    case "ovpFlow":
      setOvpQrData("")
    case "credentialOfferFlow":
      setCredentialOfferData("")
    default:
      break
    }
  }
  
}
